### PR TITLE
Engine as visitor

### PR DESCRIPTION
**Pull request by mdnazmulkarim**

### CODE REVIEW

**Review by brynrhodes — CHANGES_REQUESTED** [review]
This looks fantastic, nice work! I like the rename of runtime context to State, and I like the introduction of the Environment. The only comment I have is that we should update the comment on the CqlEngine class to document that we switched from a deserialization approach to a visitor approach, since the documentation still says we're opting for the deserialization approach.

File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/visiting/FunctionRefEvaluator.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import org.hl7.elm.r1.FunctionRef;
+import org.opencds.cqf.cql.engine.execution.CqlEngine;
+import org.opencds.cqf.cql.engine.execution.State;
+import org.opencds.cqf.cql.engine.execution.Variable;
+
+import java.util.ArrayList;
+
+public class FunctionRefEvaluator {
+
+    private FunctionDef cachedFunctionDef;
```
Comment (JPercival) [inline Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/visiting/FunctionRefEvaluator.java:14]: Making this an instance breaks the function caching. We should make this class static, and add a  `Map<FunctionRef, FunctionDef> functionCache` the `State` object. Then this evaluator should be able to use/update the state as needed to resolve the cache.

File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/visiting/QueryEvaluator.java
Comment (JPercival) [inline Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/visiting/QueryEvaluator.java:1]: This class can be used statically.

File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/visiting/ExpressionDefEvaluator.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+            if (state.getCache().isExpressionCachingEnabled()) {
+                var er = new ExpressionResult(value, state.getEvaluatedResources());
+                state.getCache().cacheExpression(libraryId, expressionDef.getName(), er);
+            }
+
+            return value;
+
+        } catch (Exception e) {
+            visitor.processException(e, expressionDef);
```
Comment (JPercival) [inline Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/visiting/ExpressionDefEvaluator.java:33]: `processException` sneakily throws an exception. We should make that clear by calling it something like `throwExceptionWithSourceLocators` or by doing something like "throw addSourceLocatorsToException(e)"

Additionally, we should centralize the exception handling. It should only be done at the top-level `evaluate` method in the cql engine.

File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/execution/CqlEngine.java [hunk 3]
```diff
@@ -315,4 +372,1416 @@ private Set<String> getExpressionSet(Library library) {
...
+    }
+
+    @Override
+    public Object visitExpressionDef(ExpressionDef expressionDef, State state) {
+
+        //logger.info(String.format("expression def visit: %s", expressionDef.getName()));
+        return ExpressionDefEvaluator.internalEvaluate(expressionDef, state, this);
+    }
+
+    public Object validateOperand(Object operand) {
```
Comment (JPercival) [inline Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/execution/CqlEngine.java:404]: It's unclear to me why we need this. Shouldn't `ExpressionDefs` be visited already when resolving the operand?

File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/execution/CqlEngine.java [hunk 4]
```diff
@@ -315,4 +372,1416 @@ private Set<String> getExpressionSet(Library library) {
...
+    }
+
+    @Override
+    public Object visitGreater(Greater elm, State state) {
+        try {
+            Object left = validateOperand(visitExpression(elm.getOperand().get(0), state));
+            Object right = validateOperand(visitExpression(elm.getOperand().get(1), state));
+
+            return GreaterEvaluator.greater(left, right, state);
+        } catch (Exception e) {
```
Comment (JPercival) [inline Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/execution/CqlEngine.java:575]: we should move exception handling up to a central location. All the visit methods should be trivial.

File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/visiting/FunctionRefEvaluator.java (hunk 1 above)
Comment (JPercival) [reply Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/visiting/FunctionRefEvaluator.java:14]: I mean the `Cache` object, not the `State` object.

**Comment by JPercival** [comment]
Approved. Let's address the remaining issues as part of new PRs.

**Review by brynrhodes — APPROVED** [review]
Overall it looks really good and I'm comfortable it's at least as functional as the previous engine based on regressions

**Pull request merged**